### PR TITLE
Add Cereal serialization and compressed SpaceTrajectory

### PR DESCRIPTION
**Pull request by mlund**
This add support for [Cereal serialisation](https://uscilab.github.io/cereal/) used to implement a zlib compressed trajectory format storing all information from `Space` (geometry is still missing, though). Related to #196.

### CODE REVIEW

**Comment by rc83** [comment]
I am used to always put a compound statement (`{}`) after `if`, `else`, `for`, etc., even if only a single statement follows. It makes later addition of new statements easier, especially if aforementioned constructs are nested.

File: src/analysis.h [hunk 1]
```diff
@@ -544,6 +548,34 @@ class QRtraj : public Analysisbase {
...
+ *
+ * @todo Geometry information
+ */
+class SpaceTrajectory : public Analysisbase {
+  private:
+    Space::Tgvec &groups; // reference to all groups
+    std::string filename;
+    std::unique_ptr<std::ostream> stream;
+    std::unique_ptr<cereal::BinaryOutputArchive> archive;
+    bool compression = true;
```
[the remark this answers is not in the document]
Comment (mlund) [reply src/analysis.h]: good point, done

File: src/analysis.cpp [hunk 2]
```diff
@@ -1435,5 +1439,43 @@ void VirtualTranslate::_to_disk() {
...
+    from_json(j);
+    name = "space trajectory";
+    filename = j.at("file");
+    std::string suffix = filename.substr(filename.find_last_of(".") + 1);
+    if (suffix == "ztraj")
+        compression = true;
+    else if (suffix == "traj")
+        compression = false;
+    else
+        throw std::runtime_error("Trajectory file suffix must be `.traj` or `.ztraj`");
```
[the remark this answers is not in the document]
Comment (mlund) [reply src/analysis.cpp]: done

File: src/analysis.cpp [hunk 3]
```diff
@@ -1435,5 +1439,43 @@ void VirtualTranslate::_to_disk() {
...
+}
+
+void SpaceTrajectory::_sample() {
+    assert(archive);
+    for (auto &group : groups) {
+        (*archive)(group);
+    }
+}
+
+void SpaceTrajectory::_to_json(json &j) const { j = {{"file", filename}, {"compress", compression}}; }
```
[the remark this answers is not in the document]
Comment (mlund) [reply src/analysis.cpp]: done

File: src/geometry.h [hunk 4]
```diff
@@ -236,6 +259,11 @@ class HexagonalPrism : public GeometryImplementation {
     std::unique_ptr<GeometryImplementation> clone() const override {
         return std::make_unique<HexagonalPrism>(*this);
     }; //!< A unique pointer to a copy of self.
+
+    //!< Cereal serialisation
+    template <class Archive> void serialize(Archive &archive) {
+        archive(cereal::base_class<GeometryImplementation>(this), rhombic2cartesian, cartesian2rhombic, box);
```
[the remark this answers is not in the document]
Comment (mlund) [reply src/geometry.h]: done

File: src/geometry_test.h [hunk 5]
```diff
@@ -367,6 +368,28 @@ TEST_CASE("[Faunus] Chameleon") {
...
+            // Chameleon chameleon(geo, CUBOID);
+            std::ifstream is("out.cereal", std::ios::binary);
+            cereal::BinaryInputArchive archive(is);
+            archive(geo);
+            CHECK(geo.getLength().x() == Approx(x));
+            CHECK(geo.getLength().y() == Approx(y));
+            CHECK(geo.getLength().z() == Approx(z));
+        }
+    }
 }
```
[the remark this answers is not in the document]
Comment (mlund) [reply src/geometry_test.h]: agree, and did do this in the group test; copied the solution here.

File: src/group.h [hunk 6]
```diff
@@ -187,4 +187,30 @@ namespace Faunus {
...
+
+template <class Archive, class T> void save(Archive &archive, const Group<T> &g) {
+    archive(g.id, g.confid, g.cm, g.compressible, g.atomic, g.size(), g.capacity());
+    for (auto it = g.begin(); it != g.trueend(); it++)
+        archive(*it);
+} //!< Cereal serialisation
+
+template <class Archive, class T> void load(Archive &archive, Group<T> &g) {
+    size_t size = 0, capacity = 0;
+    archive(g.id, g.confid, g.cm, g.compressible, g.atomic, size, capacity);
```
[the remark this answers is not in the document]
Comment (mlund) [reply src/group.h]: Yes, Cereal can do versioning by adding an extra argument to the serialization function. I have marked the Trajectory format as "experimental" in the documentation and we should handle this before release

File: src/geometry.h [hunk 7]
```diff
@@ -95,8 +100,11 @@ class GeometryImplementation : public GeometryBase {
 
     virtual ~GeometryImplementation();
 
-    //! A unique pointer to a copy of self. To be used in copy constructors.
```
[the remark this answers is not in the document]
Comment (mlund) [reply src/geometry.h]: fixed

**Pull request merged**